src/analogue.c

```diff
@@ -0,0 +1,126 @@
+#include <stdint.h>
+#include <math.h>
+
+#include <libopencm3/stm32/adc.h>
+#include <libopencm3/stm32/rcc.h>
+#include <libopencm3/stm32/gpio.h>
+#include <libopencm3/cm3/nvic.h>
+#include <libopencm3/stm32/timer.h>
+#include <libopencm3/stm32/dbgmcu.h>
+
+uint16_t input_voltage = 0;
+
+static void init_adc_timer(void) {
+    rcc_periph_clock_enable(RCC_TIM1);
+
+    timer_set_mode(TIM1, TIM_CR1_CKD_CK_INT, TIM_CR1_CMS_EDGE, TIM_CR1_DIR_UP);
+    // 24MHz/12kHz = 2kHz timer clock. /20 gives 100Hz sampling frequency
+    timer_set_period(TIM1, 12000);
+    timer_set_prescaler(TIM1, 20);
+
+    // Configure OC1 to generate an event on match
+    timer_set_oc_mode(TIM1, TIM_OC1, TIM_OCM_PWM1);
+    timer_enable_oc_preload(TIM1, TIM_OC1);
+    timer_set_oc_value(TIM1, TIM_OC1, 1);
+
+    // Enable trigger output for triggering ADC scan run
+    timer_set_master_mode(TIM1, TIM_CR2_MMS_COMPARE_OC1REF);
+
+    timer_enable_preload(TIM1);
+
+    /* Halt the ADC timer while debugging */
+    DBGMCU_CR |= DBGMCU_CR_TIM1_STOP;
+}
+
+static void init_adc(void) {
+    gpio_set_mode(GPIOB, GPIO_MODE_INPUT, GPIO_CNF_INPUT_ANALOG, GPIO1);  // 12V
+    gpio_set_mode(GPIOC, GPIO_MODE_INPUT, GPIO_CNF_INPUT_ANALOG, GPIO3);  // M0 CS
+    gpio_set_mode(GPIOC, GPIO_MODE_INPUT, GPIO_CNF_INPUT_ANALOG, GPIO0);  // M1 CS
+
+    rcc_periph_clock_enable(RCC_ADC1);
+    rcc_set_adcpre(RCC_CFGR_ADCPRE_PCLK2_DIV2);
+
+    // Disable ADC during configuration
+    adc_power_off(ADC1);
+
+    // Extend the conversion times to reduce noise impact
+    adc_set_sample_time_on_all_channels(ADC1, ADC_SMPR_SMP_28DOT5CYC);
+    adc_set_right_aligned(ADC1);
+
+    // Enable an interrupt after each scan run
+    nvic_enable_irq(NVIC_ADC1_2_IRQ);
+    nvic_set_priority(NVIC_ADC1_2_IRQ, 1);
+    adc_enable_eoc_interrupt(ADC1);
+
+    // Enable scan runs on timer 1 trigger output
+    adc_enable_external_trigger_injected(ADC1, ADC_CR2_JEXTSEL_TIM1_TRGO);
+    adc_enable_scan_mode(ADC1);
+
+    // Configure the channels to be sampled in each scan run
+    // the outputs will be in the ADC_JDRx registers
+    uint8_t channel[] = {9, 13, 10};
+    adc_set_injected_sequence(ADC1, 3, channel);
+
+    adc_power_on(ADC1);
+
+    // Wait >1.3us for ADC to be ready to perform conversions
+    __asm__(  // wait 40 clock cycles @ 24MHz = ~1.6us
+        "nop;nop;nop;nop;nop;"
+        "nop;nop;nop;nop;nop;"  // 10
+        "nop;nop;nop;nop;nop;"
+        "nop;nop;nop;nop;nop;"  // 20
+        "nop;nop;nop;nop;nop;"
+        "nop;nop;nop;nop;nop;"  // 30
+        "nop;nop;nop;nop;nop;"
+        "nop;nop;nop;nop;nop;"  // 40
+    );
+    adc_reset_calibration(ADC1);
+    adc_calibrate(ADC1);
+}
+
+void analogue_init(void) {
+    init_adc_timer();
+    init_adc();
+
+    timer_enable_counter(TIM1);
+}
+
+static uint16_t convert_to_ma(uint16_t current_raw) {
+    // voltage_mv = code * vref/4096
+    // current_ma = (voltage_mv/rshunt) * Igain
+
+    // current_ma = code * 3300/(4096 * 1100) * 7000
```

Comment:
question: Is this pseudocode for line 99? Please could you clarify in the comment if so.

Reply:
This is the electronics equation to resolve the ADC code to mA.